File: benchmarks/R/benchmark_synthdid.R
#!/usr/bin/env Rscript
# Benchmark: Synthetic DiD (R `synthdid` package)
#
# Usage:
#   Rscript benchmark_synthdid.R --data path/to/data.csv --output path/to/results.json

library(synthdid)
library(jsonlite)
library(data.table)

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

parse_bool <- function(x, flag) {
  v <- tolower(x)
  if (v %in% c("true", "t", "1", "yes")) return(TRUE)
  if (v %in% c("false", "f", "0", "no")) return(FALSE)
  stop(sprintf("Invalid boolean for %s: '%s' (use true/false)", flag, x))
}

parse_args <- function(args) {
  result <- list(
    data = NULL,
    output = NULL,
    warmup = FALSE,
    skip_jackknife = FALSE
  )

  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data") {
      result$data <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--output") {
      result$output <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--warmup") {
      result$warmup <- parse_bool(args[i + 1], "--warmup")
      i <- i + 2
    } else if (args[i] == "--skip-jackknife") {
      result$skip_jackknife <- parse_bool(args[i + 1], "--skip-jackknife")
      i <- i + 2
    } else {
      # Unknown flags used to be silently skipped, which turned typos into
      # silent protocol changes. Fail loudly instead.
      stop(sprintf("Unknown flag: %s", args[i]))
    }
  }

  if (is.null(result$data) || is.null(result$output)) {
    stop("Usage: Rscript benchmark_synthdid.R --data <path> --output <path>")
  }

  return(result)
}

config <- parse_args(args)

# Load data
message(sprintf("Loading data from: %s", config$data))
data <- fread(config$data)

# synthdid requires panel.matrices format
# Data must have: unit, time, outcome, treated columns
message("Preparing data for synthdid...")

# Create treatment indicator (1 if treated in post period)
# synthdid expects 0/1 treatment indicator
data[, treatment := as.integer(treated == 1 & post == 1)]

# Convert to panel.matrices format
setup <- panel.matrices(
  as.data.frame(data),
  unit = "unit",
  time = "time",
  outcome = "outcome",
  treatment = "treatment"
)

# Warm-up: run the FULL timed pipeline (estimation + placebo vcov) once
# untimed so byte-compiler JIT and first-call setup stay out of the window.
if (config$warmup) {
  message("Warm-up fit (untimed)...")
  tau_warm <- synthdid_estimate(setup$Y, setup$N0, setup$T0)
  invisible(vcov(tau_warm, method = "placebo"))
  rm(tau_warm)
}

# Run benchmark
message("Running Synthetic DiD estimation...")
start_time <- Sys.time()

tau_hat <- synthdid_estimate(setup$Y, setup$N0, setup$T0)

estimation_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

# Get weights
weights <- attr(tau_hat, "weights")
unit_weights <- weights$omega
time_weights <- weights$lambda

# Compute SE via placebo
message("Computing standard errors...")
se_start <- Sys.time()
se_matrix <- vcov(tau_hat, method = "placebo")
se <- as.numeric(sqrt(se_matrix[1, 1]))  # Extract scalar SE
se_time <- as.numeric(difftime(Sys.time(), se_start, units = "secs"))

# Compute SE via jackknife (Algorithm 3). Not part of total_seconds (which
# is estimation + placebo SE, matching the Python arm); skippable because it
# wastes hours of wall-clock at large scales in timing-focused runs.
if (config$skip_jackknife) {
  message("Skipping jackknife standard errors (--skip-jackknife)")
  se_jackknife <- NA_real_
  se_jk_time <- NA_real_
} else {
  message("Computing jackknife standard errors...")
  se_jk_start <- Sys.time()
  se_jk_matrix <- vcov(tau_hat, method = "jackknife")
  se_jackknife <- as.numeric(sqrt(se_jk_matrix[1, 1]))
  se_jk_time <- as.numeric(difftime(Sys.time(), se_jk_start, units = "secs"))
}

total_time <- estimation_time + se_time  # placebo only, matches `se` field

# Compute noise level and regularization (to match Python's auto-computed values)
N0 <- setup$N0
T0 <- setup$T0
N1 <- nrow(setup$Y) - N0
T1 <- ncol(setup$Y) - T0
noise_level <- sd(apply(setup$Y[1:N0, 1:T0], 1, diff))
zeta_omega <- ((N1 * T1)^(1/4)) * noise_level
zeta_lambda <- 1e-6 * noise_level

# Format output
results <- list(
  estimator = "synthdid::synthdid_estimate",

  # Point estimate and SE
  att = as.numeric(tau_hat),
  se = se,
  se_jackknife = se_jackknife,

  # Weights (full precision; omega is in panel control-row order, lambda in
  # pre-period column order - emit the ids so comparison aligns by key)
  unit_weights = as.numeric(unit_weights),
  unit_weight_ids = rownames(setup$Y)[seq_len(N0)],
  time_weights = as.numeric(time_weights),
  time_weight_ids = colnames(setup$Y)[seq_len(T0)],

  # Regularization parameters
  noise_level = noise_level,
  zeta_omega = zeta_omega,
  zeta_lambda = zeta_lambda,

  # Timing
  timing = list(
    estimation_seconds = estimation_time,
    se_placebo_seconds = se_time,
    se_jackknife_seconds = se_jk_time,
    total_seconds = total_time
  ),

  # Metadata
  metadata = list(
    r_version = R.version.string,
    synthdid_version = as.character(packageVersion("synthdid")),
    dt_threads = data.table::getDTthreads(),
    warmup = config$warmup,
    skip_jackknife = config$skip_jackknife,
    n_control = N0,
    n_treated = N1,
    n_pre_periods = T0,
    n_post_periods = T1
  )
)

# Write output
message(sprintf("Writing results to: %s", config$output))
write_json(results, config$output, auto_unbox = TRUE, pretty = TRUE, digits = 10)

message(sprintf("Completed in %.3f seconds", total_time))

File: benchmarks/python/benchmark_synthdid.py
#!/usr/bin/env python3
"""
Benchmark: Synthetic DiD (diff-diff SyntheticDiD).

Usage:
    python benchmark_synthdid.py --data path/to/data.csv --output path/to/results.json
"""

import argparse
import json
import sys
from pathlib import Path

import numpy as np
import pandas as pd

# Add parent to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from diff_diff import SyntheticDiD
from benchmarks.python.utils import Timer


def parse_args():
    parser = argparse.ArgumentParser(description="Benchmark Synthetic DiD estimator")
    parser.add_argument("--data", required=True, help="Path to input CSV data")
    parser.add_argument("--output", required=True, help="Path to output JSON results")
    parser.add_argument(
        "--n-bootstrap", type=int, default=200, help="Number of bootstrap iterations"
    )
    parser.add_argument(
        "--variance-method", type=str, default="placebo",
        choices=["bootstrap", "placebo"],
        help="Variance estimation method (default: placebo to match R)"
    )
    parser.add_argument(
        "--backend", default="auto", choices=["auto", "python", "rust"],
        help="Backend to use: auto (default), python (pure Python), rust (Rust backend)"
    )
    return parser.parse_args()


def configure_backend(backend: str) -> str:
    """Configure the backend and return the actual backend being used."""
    import diff_diff

    if backend == "python":
        # Force pure Python by disabling Rust backend
        diff_diff.HAS_RUST_BACKEND = False
        diff_diff._rust_solve_ols = None
        diff_diff._rust_compute_robust_vcov = None
        diff_diff._rust_bootstrap_weights = None
        diff_diff._rust_synthetic_weights = None
        diff_diff._rust_project_simplex = None
        return "python"
    elif backend == "rust":
        if not diff_diff.HAS_RUST_BACKEND:
            raise RuntimeError("Rust backend requested but not available")
        return "rust"
    else:  # auto
        return "rust" if diff_diff.HAS_RUST_BACKEND else "python"


def main():
    args = parse_args()

    # Configure backend before running estimation
    actual_backend = configure_backend(args.backend)
    print(f"Using backend: {actual_backend}")

    # Load data
    print(f"Loading data from: {args.data}")
    data = pd.read_csv(args.data)

    # Determine post periods from data
    # Assumes 'post' column exists with 0/1 indicators
    post_periods = sorted(data[data["post"] == 1]["time"].unique().tolist())

    # Run benchmark
    print("Running Synthetic DiD estimation...")
    sdid = SyntheticDiD(
        n_bootstrap=args.n_bootstrap,
        variance_method=args.variance_method,
        seed=42
    )

    with Timer() as timer:
        results = sdid.fit(
            data,
            outcome="outcome",
            treatment="treated",
            unit="unit",
            time="time",
            post_periods=post_periods,
        )

    total_time = timer.elapsed

    # Get weights
    unit_weights_df = results.get_unit_weights_df()
    time_weights_df = results.get_time_weights_df()

    # Build output
    output = {
        "estimator": "diff_diff.SyntheticDiD",
        "backend": actual_backend,
        # Point estimate and SE
        "att": float(results.att),
        "se": float(results.se),
        # Weights
        "unit_weights": unit_weights_df["weight"].tolist(),
        "time_weights": time_weights_df["weight"].tolist(),
        # Timing
        "timing": {
            "estimation_seconds": total_time,
            "total_seconds": total_time,
        },
        # Metadata
        "metadata": {
            "n_control": len(data[data["treated"] == 0]["unit"].unique()),
            "n_treated": len(data[data["treated"] == 1]["unit"].unique()),
            "n_pre_periods": len(data[data["post"] == 0]["time"].unique()),
            "n_post_periods": len(post_periods),
            "n_bootstrap": args.n_bootstrap,
            "variance_method": args.variance_method,
        },
    }

    # Write output
    print(f"Writing results to: {args.output}")
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, "w") as f:
        json.dump(output, f, indent=2)

    print(f"Completed in {total_time:.3f} seconds")
    return output


if __name__ == "__main__":
    main()
